builtin/logical/aws/secret_access_keys.go

```diff
@@ -43,26 +47,45 @@ func secretAccessKeys(b *backend) *framework.Secret {
 	}
 }
 
-func genUsername(displayName, policyName, userType string) (ret string, warning string) {
-	var midString string
+func genUsername(displayName, policyName, userType, usernameTemplate string) (ret string, warning string) {
 
 	switch userType {
 	case "iam_user":
 		// IAM users are capped at 64 chars; this leaves, after the beginning and
 		// end added below, 42 chars to play with.
-		midString = fmt.Sprintf("%s-%s-",
-			normalizeDisplayName(displayName),
-			normalizeDisplayName(policyName))
-		if len(midString) > 42 {
-			midString = midString[0:42]
+		up, err := template.NewTemplate(template.Template(usernameTemplate))
+		if err != nil {
+			return "", fmt.Sprintf("unable to initialize username template: %s", err)
```

Comment:
If the template instantiation fails, should this be a fatal error instead of a warning? Similarly, on the `up.Generate(um)` call below the error should be propagated all the way to the caller since that will result in an empty username. We should consider changing `genUsername` to return an error.

Reply:
I agree with that! I wasn't 100% on if it should be an error because of the IAM length capping warning, but it makes sense